src/web_interface/rest/rest_firmware.py

```diff
@@ -18,68 +18,40 @@
-        firmware_object = convert_analysis_task_to_fw_obj(data)
-        with ConnectTo(InterComFrontEndBinding, self.config) as intercom:
-            intercom.add_analysis_task(firmware_object)
-        data.pop('binary')
+api = Namespace('rest/firmware', description='Query the firmware database or upload a firmware')
 
-        return dict(uid=firmware_object.uid)
 
-    @staticmethod
-    def _fit_firmware(firmware):
-        meta = create_meta_dict(firmware)
-        analysis = firmware.processed_analysis
-        return dict(meta_data=meta, analysis=analysis)
+firmware_model = api.model('Upload Firmware', {
+    'device_name': fields.String(description='Device Name'),
+    'device_part': fields.String(description='Device Part', required=True),
+    'device_class': fields.String(description='Device Class'),
+    'file_name':  fields.String(description='File Name', required=True),
+    'version':  fields.String(description='Version', required=True),
+    'vendor':  fields.String(description='Vendor', required=True),
+    'release_date':  fields.String(description='Release Date'),
+    'tags':  fields.String(description='Tags'),
+    'requested_analysis_systems': fields.List(description='Analyses', cls_or_instance=fields.String),
+    'binary': fields.String(description='Base64 string representing the raw binary')
+})
 
-    def _update_analysis(self, uid, update):
-        with ConnectTo(FrontEndDbInterface, self.config) as connection:
-            firmware = connection.get_firmware(uid)
-        if not firmware:
-            return error_message('No firmware with UID {} found'.format(uid), self.URL, dict(uid=uid))
 
-        unpack = 'unpacker' in update
-        while 'unpacker' in update:
-            update.remove('unpacker')
+class RestFirmwareBase(Resource):
+    URL = '/rest/firmware'
 
-        firmware.scheduled_analysis = update
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.config = kwargs.get('config', None)
 
-        with ConnectTo(InterComFrontEndBinding, self.config) as intercom:
-            supported_plugins = intercom.get_available_analysis_plugins().keys()
-            for item in update:
-                if item not in supported_plugins:
-                    return error_message('Unknown analysis system \'{}\''.format(item), self.URL, dict(uid=uid, update=update))
-            intercom.add_re_analyze_task(firmware, unpack)
 
-        if unpack:
-            update.append('unpacker')
-        return success_message({}, self.URL, dict(uid=uid, update=update))
-
-    def _get_without_uid(self):
+@api.route('', doc={'description': ''})
+class RestFirmwareGetWithoutUid(RestFirmwareBase):
+    @roles_accepted(*PRIVILEGES['view_analysis'])
+    @api.doc(responses={200: 'Success', 400: 'Unknown file object'})
+    def get(self):
+        '''
+        Browse the database
+        List all available firmwares in the database
```

Comment:
firmware is not countable (like software): firmwares -> firmware

Reply:
Documentation of optional URL arguments (offset, limit, query, recursive and inverted) is missing (see https://github.com/fkie-cad/FACT_core/wiki/Rest-API#get)